### DISCUSSION TITLE
How to draw wrapped text?

### DISCUSSION THREAD

**Discussion by AshfordN** [fltk-rs/fltk-rs#896] [Q&A] [answered]
I am trying to display lines of text as a bulleted list. So far, I've explored the possibility of using a `TextDisplay`, but there is no way to indent the wrapped lines so that they line up with the first line. Instead, the text ends up looking like the following:
![rect1241](https://user-images.githubusercontent.com/4706674/134034681-948d2ee5-4ddc-4b38-894a-b1d368f69eb2.png)

I'm also considering the possibility of drawing the text on a frame using the draw functions, but there doesn't appear to be an easy way to achieve text wrapping using this method. The closest I've seen, is calling `draw_text2` with `Align::Wrap` for the `align` argument, but that only clips the text rather than wrap it. So the text ends up looking like the following:
![rect1563](https://user-images.githubusercontent.com/4706674/134035691-d5fb7146-7ab6-4512-ab4f-45a1001342a5.png)

What's the best way to achieve what I'm asking?

**Comment by MoAlyousef** [comment]
FLTK's text widgets only support plain text, minimal formatting (color, size and font) can be done using style buffers, more special formatting will have to be done in user code or via some external library. 

Though the HelpView widget supports HTML 2.0 elements, including lists and their enclosed block text:

![image](https://user-images.githubusercontent.com/37966791/134044633-cc6b659c-2571-444a-9bdb-7edbf5c114d9.png)

```rust
use fltk::{
    app, 
    misc,
    prelude::*,
    window,
};

const TEXT: &str = r#"
<ul>
<li>This text is <font color=red>long</font> and should wrap</li>
<p><!-- just inserting a space between list elements! -->
<li>This text is also very long and should wrap</li>
</ul>
"#;

fn main() {
    let app = app::App::default();
    let mut wind = window::Window::default().with_size(400, 300);
    let mut view = misc::HelpView::default().with_size(390, 290).center_of_parent();
    wind.end();
    wind.show();
    view.set_text_size(18);
    view.set_value(TEXT); // doesn't need to be const or static!

    app.run().unwrap();
}
```

There's also [fltk-webview](https://crates.io/crates/fltk-webview) which embeds a webview inside an fltk app, that repo has an example using markdown:
https://github.com/fltk-rs/fltk-webview/blob/main/examples/markdown.rs

**Comment by AshfordN** [comment, marked as answer]
The `HelpView` widget is a somewhat practical option, but the default layout (and styling) seems a bit awkward for what I'm doing. Specifically, the indentation of the bulleted list is too large; and there's no way to adjust it. I would probably be able to avoid this limitation by using [fltk-webview](https://crates.io/crates/fltk-webview), but that feels like an over-engineered solution for what I'm doing. Instead, I've taken the first approach by implementing basic text wrapping functionality, as shown below:
```rust
use fltk::draw;
use unicode_linebreak::linebreaks;
use unicode_linebreak::BreakOpportunity::Mandatory;

pub fn draw_wrapped_text(string: &str, x: i32, y: i32, w: i32) -> i32 {
    let line_height = draw::height();
    let mut y = y;
    let mut start_idx = 0;
    let mut end_idx = 0;
    let mut progress = 0;
    for (idx, opp) in linebreaks(string) {
        let (width, _) = draw::measure(&string[start_idx..idx], false);
        if width > w {
            draw::draw_text(&string[start_idx..end_idx], x, y);
            y += line_height;
            start_idx = end_idx;
            progress = end_idx;
        } else if opp == Mandatory {
            draw::draw_text(&string[start_idx..idx], x, y);
            y += line_height;
            start_idx = idx;
            progress = idx;
        }

        end_idx = idx;
    }

    if progress < end_idx {
        draw::draw_text(&string[start_idx..end_idx], x, y);
        y += line_height;
    }

    // return the y coordinate of the bottom of the wrapped text
    y
}
```
This, of course, over-simplifies the text wrapping problem, and does not cover the entire problem domain; but, it suffices for what I need.
